File: 2020/day/7/Main.hs
import Data.Char (isAlpha, isDigit)
import Data.Maybe (fromMaybe)
import Text.ParserCombinators.ReadP

type ColorCode = String

type Graph = [(ColorCode, [(ColorCode, Integer)])]

colorCode :: ReadP ColorCode
colorCode = do
    adjective <- munch1 isAlpha
    char ' '
    color <- munch1 isAlpha
    return (adjective ++ ' ' : color)

bags :: ReadP (ColorCode, Integer)
bags = do
    n <- read <$> munch1 isDigit
    char ' '
    key <- colorCode
    if n == 1
        then string " bag"
        else string " bags"
    return (key, n)

parseLine :: ReadP (ColorCode, [(ColorCode, Integer)])
parseLine = do
    bag <- colorCode
    string " bags contain "
    contents <- 
        (string "no other bags" *> pure [])
        +++
        (bags `sepBy` string ", ")
    char '.'
    return (bag, contents)

parse :: ReadP Graph
parse = parseLine `endBy` char '\n' <* eof

-- Returns list of vertices reachable from a given vertex.
reachable :: Graph -> ColorCode -> [ColorCode]
reachable g v = v : maybe [] (concatMap (reachable g . fst)) (lookup v g)

countBagsInside :: Graph -> ColorCode -> Integer
countBagsInside g v = sum [n + n * countBagsInside g w | (w, n) <- fromMaybe [] (lookup v g)]

main = do
    input <- readFile "input.txt"
    let [(graph, "")] = readP_to_S parse input

    -- Part 1
    print $ length $ filter (elem "shiny gold" . tail . reachable graph . fst) graph

    -- Part 2
    print $ countBagsInside graph "shiny gold"

File: 2020/day/7/main.py
from collections import deque
import fileinput

G = {}

for line in fileinput.input('input.txt'):
    line = line.strip()
    words = line.split()
    key = ' '.join(words[0:2])
    assert words[2] == 'bags'
    assert words[3] == 'contain'

    for i in range(4, len(words), 4):
        if words[i] == 'no':
            n = 0
            break
        else:
            n = int(words[i])

        key2 = ' '.join(words[i+1:i+3])
        assert words[i + 3] in ['bag,', 'bags,', 'bag.', 'bags.'], words[i + 3]
        G.setdefault(key, set()).add((key2, n))

# Part 1
# All nodes that are connected to shiny gold
ans1 = 0

def reachable(x, G):
    yield x
    for y, m in G.get(x, []):
        for z in reachable(y, G):
            yield z

for x in G:
    if x != 'shiny gold':
        ans1 += 'shiny gold' in reachable(x, G)

print(ans1)

# Part 2
# Total number of bags inside the shiny gold bag
def part2(x):
    return sum(n + n * part2(y) for y, n in G.get(x, []))

print(part2('shiny gold'))